odoo/addons/base/ir/ir_actions_report.py

```diff
@@ -576,19 +483,89 @@ def render_template(self, template, values=None):
         return view_obj.render_template(template, values)
 
     @api.model
-    def render_qweb_pdf(self, res_ids, html=None, data=None):
+    def _post_pdf(self, attachment_ids, pdf_content=None, res_ids=None):
+        '''Merge the existing attachments by adding one by one the content of the attachments
+        and then, we add the pdf_content if exists. Create the attachments for each record individually
+        if required.
+
+        :param attachment_ids: The retrieved attachments.
+        :param pdf_content: The pdf content newly generated by wkhtmltopdf.
+        :param res_ids: the ids of record to allow postprocessing.
+        :return: The pdf content of the merged pdf.
+        '''
+
+        def close_streams(streams):
+            for stream in streams:
+                try:
+                    stream.close()
+                except Exception:
+                    pass
+
+        # Check special case having only one record with existing attachment.
+        if len(attachment_ids) == 1 and not pdf_content:
+            return base64.decodestring(attachment_ids[0].datas)
+
+        # Append all attachments content and then, append pdf_content
+        writer = PdfFileWriter()
+        streams = []
+        # Decode each attachment and append
+        for attachment_id in attachment_ids:
+            content = base64.decodestring(attachment_id.datas)
+            stream = StringIO(content)
+            streams.append(stream)
+            reader = PdfFileReader(stream)
+            writer.appendPagesFromReader(reader)
+        # Append content of pdf if exists
+        if pdf_content:
+            stream = StringIO(pdf_content)
+            streams.append(stream)
+            reader = PdfFileReader(stream)
+            # Call the postprocess method for each record.
+            if res_ids and self.attachment_use and self.attachment:
+                # Build a record_map mapping id -> record
+                record_map = dict(
+                    [(r.id, r) for r in self.env[self.model].browse([res_id for res_id in res_ids if res_id])])
+                if len(res_ids) == 1 and res_ids[0] in record_map:
+                    # Only one record, so postprocess directly and append the whole pdf.
+                    self.postprocess_pdf_report(record_map[res_ids[0]], pdf_content)
```

Comment:
@smetl Yes it's nice to have wkhtmltopdb called only once. By reading your code I've the feeling it's possible to adapt this code to put into the final result a StringIO that could be passed to the postprocess method without additional performance cost.

```python
      # Append content of pdf if exists
        if pdf_content:
            content_streams = []
            # Call the postprocess method for each record.
            if res_ids and self.attachment_use and self.attachment:
                # Build a record_map mapping id -> record
                record_map = dict(
                    [(r.id, r) for r in self.env[self.model].browse([res_id for res_id in res_ids if res_id])])
                if len(res_ids) == 1 and res_ids[0] in record_map:
                    # Only one record, so postprocess directly and append the whole pdf.
                    stream = StringIO(pdf_content)
                    self.postprocess_pdf_report(record_map[res_ids[0]], stream)
                    content_streams.append(streams)
                else:
                    # In case of multiple docs, we need to split the pdf according the records.
                    # To do so, we split the pdf based on outlines computed by wkhtmltopdf.
                    # An outline is a <h?> html tag found on the document. To retrieve this table,
                    # we look on the pdf structure using pypdf to compute the outlines_pages that is
                    # an array like [0, 3, 5] that means a new document start at page 0, 3 and 5.
                    outlines_pages = sorted(
                        [outline.getObject()[0] for outline in reader.trailer['/Root']['/Dests'].values()])
                    assert len(outlines_pages) == len(res_ids)
                    for i, num in enumerate(outlines_pages):
                        if not res_ids[i]:
                            continue
                        to = outlines_pages[i + 1] if i + 1 < len(outlines_pages) else reader.numPages
                        writer = PdfFileWriter()
                        for j in range(num, to):
                            writer.addPage(reader.getPage(j))
                        attachment_content = StringIO()
                        writer.write(attachment_content)
                        self.postprocess_pdf_report(record_map[res_ids[i]], attachment_content)
                        content_streams.append(attachment_content)
            else:
                stream = StringIO(pdf_content)
                content_streams.append(stream)

            # add content to the result
            for stream in content_streams:
                reader = PdfFileReader(stream)
                writer.appendPagesFromReader(reader)
            close_streams(content_streams)
```